hoot-core/src/main/cpp/hoot/core/visitors/FindIntersectionsVisitor.cpp

```diff
@@ -24,42 +24,46 @@
  *
  * @copyright Copyright (C) 2015, 2017, 2018, 2019 DigitalGlobe (http://www.digitalglobe.com/)
  */
-#include "FindHighwayIntersectionsVisitor.h"
+#include "FindIntersectionsVisitor.h"
 
+#include <hoot/core/conflate/matching/NodeMatcher.h>
+#include <hoot/core/criterion/HighwayCriterion.h>
+#include <hoot/core/criterion/RailwayCriterion.h>
+#include <hoot/core/elements/NodeToWayMap.h>
 #include <hoot/core/elements/OsmMap.h>
 #include <hoot/core/index/OsmMapIndex.h>
-#include <hoot/core/elements/NodeToWayMap.h>
-#include <hoot/core/conflate/matching/NodeMatcher.h>
 #include <hoot/core/util/Factory.h>
-#include <hoot/core/criterion/HighwayCriterion.h>
 
 using namespace std;
 
 namespace hoot
 {
 
 HOOT_FACTORY_REGISTER(ElementVisitor, FindHighwayIntersectionsVisitor)
+HOOT_FACTORY_REGISTER(ElementVisitor, FindRailwayIntersectionsVisitor)
 
-void FindHighwayIntersectionsVisitor::visit(const ConstElementPtr& e)
+void FindIntersectionsVisitor::visit(const ConstElementPtr& e)
 {
   std::shared_ptr<NodeToWayMap> n2w = _map->getIndex().getNodeToWayMap();
   long id = e->getId();
 
   const set<long>& wids = n2w->getWaysByNode(id);
-
-  // find all ways that are highways (ie roads)
+  // create the criterion if needed
+  if (!_criterion)
+    _criterion = createCriterion(_map->shared_from_this());
```

Comment:
Create the appropriate criterion (only once might I add) in a similar fashion to the operation above.